### ISSUE TITLE
why is tensorflow.map_fn slow, what is wrong with following code?

### ISSUE THREAD

**Issue by minhhg** [tensorflow/tensorflow#24774]
I am trying to use tensorflow map_fn to do parallel computation. However it seems to me that the performance gain is not significant.

Here are example code running Python 3.6.5, Tensorflow version 1.12.0 on Ubuntu 14.04 LTS, 28 duo cores (Intel(R) Xeon(R) CPU E5-2697 v3 @ 2.60GHz) = 56 processors

These same codes running on Amazon AWS SagerMaker ml-p3-xlarge even took longer time, 227 seconds.

python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"
b'v1.12.0-0-ga6d8ffa' 1.12.0

```
import tensorflow as tf
import time
# version 1
tic = time.time()
elems = np.array(range(1,1000000), dtype=np.float64)
output = tf.map_fn(lambda x: x**6 , elems, dtype=tf.float64,  parallel_iterations=56)
sess = tf.Session()

res = sess.run(output)
toc = time.time() - tic
print("elapsed=", toc)  # 29.47 (seconds)

# version 2
tic = time.time()
elems = np.array(range(1,1000000), dtype=np.float64)
output = tf.map_fn(lambda x: x**6 , elems, dtype=tf.float64,   parallel_iterations=56)
n_cpus=28


with  tf.Session(
config=tf.ConfigProto(log_device_placement=True, 
device_count={ "CPU": n_cpus },
inter_op_parallelism_threads=n_cpus,
intra_op_parallelism_threads=1,

)) as sess:
res = sess.run(output)

toc = time.time() - tic
print("elapsed=", toc)  # 29.26 (seconds)

# version 3
tic = time.time()
elems = np.array(range(1,1000000), dtype=np.float64)
x6 = [ x**6 for x in elems]
toc = time.time() - tic
print("elapsed time=", toc) # 0.5 seconds
```

What is problem with the above codes? without map_fn, sequential execution version 3 only 0.5 (seconds).

**Comment by vidursatija** [comment]
Hi,
I think you encounter this problem because it takes time for tensorflow to create the graph and create the session. A more accurate value for elapsed time will be as follows:
```python
elems = np.array(range(1,1000000), dtype=np.float64)
output = tf.map_fn(lambda x: x**6 , elems, dtype=tf.float64, parallel_iterations=56)
sess = tf.Session()

tic = time.time()
res = sess.run(output)
toc = time.time() - tic
print("elapsed=", toc)
```

**Comment by minhhg** [comment]
Hi vidursatija,
The difference is only 1 second, I re-run with your code. It took 28 (seconds).

**Comment by minhhg** [comment]
Hi msymp,
The issue is not  solved. It is 28 (seconds) with tensorflow vs 0.5 (seconds) with python. Thanks.

**Comment by msymp** [comment]
Hi @tatianashp , can you please advise why the user is not seeing gains from tf.map_fn in parallel runs over 28 cores. Thanks.

**Comment by venbalas** [comment]
I'm having a similar issue where map_fn is the bottleneck!..

**Comment by Han-sok** [comment]
Yes. I am experiencing almost the same issue on my side. I don't want to be rude for the developers who give a lot of contribution to this tool, but I don't understand why on earth the only way to parallelize user code is so slow. Is this happen because tf.map_fn or tf.while_loop is not designed for parallelization?

**Comment by Han-sok** [comment]
I am leaving an additional comment because the former post is just grunting not suitable for asking for help. In my case, I am using multiple tf.map_fn methods within __call()__ method of my custom layer implementation using "tf.keras.layers.Layer" class. I am experiencing very slow graph construction and training time too. Since graph construction may have its own overhead, I can deal with slow graph construction. However, my custom layer is optimized for some computation. It has to be fast if tf.map_fn is really doing parallelization inside. I even optimize the code so the GPU has almost over 90 percent of utilization, but it gives only a slight speed up. The same algorithm on the custom ASIC is verified working and it's fast. So it's not related to the algorithmic problem. Could you give me any clue on this? Maybe using tf.map_fn in the custom layer is a bad idea?

**Comment by tatianashp** [comment]
@Han-sok - Can you please provide full information about your issue according to the issue template?

**Comment by GoingMyWay** [comment]
`tf.map_fn` is really slow.

**Comment by Asorie** [comment]
I experience the same problem with `tf.map_fn` being slow. Is there any way to accelerate this function?

**Comment by GoingMyWay** [comment]
> I experience the same problem with `tf.map_fn` being slow. Is there any way to accelerate this function?

For me, the best solution is to avoid using it. Use vectorization as many as possible.

**Comment by Asorie** [comment]
For my problem i have to use the `tf.boolean_mask`-function, which can produce vectors of different lenghts. This output is further reduced to a single vector. 
Thats the reason i have to use the `tf.map_fn` since it is possible to compute  the `tf.boolean_mask` for a single element, then reduce it to a scalar and when the computation is done it gets combined to a vector.

**Comment by xsr6064** [comment]
> > I experience the same problem with `tf.map_fn` being slow. Is there any way to accelerate this function?
> 
> For me, the best solution is to avoid using it. Use vectorization as many as possible.

I implemented a customized loss similar to yours, and `tf.map_fn` is running very slow as well. is it possible to share how you vectorized your computation? thanks a lot.

**Comment by Sawatdatta** [comment]
I am also facing this issue. I wrote a custom layer to process an color image. Input tensor for 16 batch size is(16, 128, 128, 32). Using tf.map_fn is taking indefinite time.

**Comment by Sawatdatta** [comment]
Please suggest a workaround

**Comment by UpCoder** [comment]
mark, I also meet this problem

**Comment by tatianashp** [comment]
@alextp - do you have any thoughts on a workaround for tf.map_fn performance?

**Comment by alextp** [comment]
Does tf.vectorized_map work here instead of map_fn?

**Comment by tejaslodaya** [comment]
I have come across the same issue as yours @Asorie.

In my case, I have features which can be variable length vectors. Because tensorflow (and matrix in general) doesn't support variable-length vectors, I am storing them as string. My map_fn converts string type 1D tensor to double 1D tensor computes a single scalar for each row (intersection) and the output of map_fn returns a 1D vector.

This is impossible to do in vectorization @GoingMyWay.

Let me know if you find a workaround @Asorie

**Comment by tejaslodaya** [comment]
@alextp I feel vectorized_map should solve this issue, but it comes with many limitations.

Will revert back if it suits my use-case

**Comment by fsx950223** [comment]
```map_fn``` will create a branch for every element, your code creates 1000000 graph branches which will cause the performance issue. You could try vectorized_map or custom op kernel.

**Comment by tejaslodaya** [comment]
@fsx950223 I wanted to change the data type as well inside the `map_fn`. This change isn't allowed in `vectorized_map`, hence the choice

**Comment by tejaslodaya** [comment]
Whoever is coming here, here are some of the pitfalls to be aware of while using `vectorized_map`: https://www.tensorflow.org/api_docs/python/tf/vectorized_map

**Comment by igormorgado** [comment]
I had the same issues. Vectorized map works faster but still subpar compared with numpy alternatives. In the same dataset size  `numpy.apply_along_axis` takes 50microseconds (over a numpy array) while `tf.vectorized_map` takes 25ms over a tensor (numpy is 500x faster), if comparing with `tf.map_fn` over a tensor, things go even worse, since it takes `300ms` (numpy is 6000x faster).

**Comment by nikhil1008** [comment]
I have same issue as well, trying to map a floor approximation function  to image pixels. Its horribly slow. The Only reason for using TF is faster parallel computation, its strange that Map_fn doesn't do that. 

![image](https://user-images.githubusercontent.com/7234284/85325721-f3f5b080-b480-11ea-9893-00ca4c78cbba.png)

**Comment by agarwal-ashish** [comment]
@nikhil1008 It looks like your code is also measuring the "graph construction cost" which is likely unintentional.

Here is my measurement, using TF 2.x:
def diff_floor(x):
  xp = 10
  a = tf.range(-xp, xp + 1, 1, 'float32')
  b = tf.tile(x, (2 * xp + 1,))
  b = b - a
  b = 5000 * b
  b = 1 + tf.math.exp(-b)
  b = 1.0 / b
  b = tf.reduce_sum(b) - xp -1
  return b

@tf.function(experimental_compile=True)
def f(a):
  return tf.map_fn(diff_floor, tf.reshape(a, (-1, 1)), parallel_iterations=8)

f(tf.constant(np.random.rand(224, 224, 3), dtype=tf.float32))  # warmup
%timeit f(tf.constant(np.random.rand(224, 224, 3), dtype=tf.float32))
 -> 10 loops, best of 3: 21.8 ms per loop

Replacing the map_fn with a vectorized_map gives:
->100 loops, best of 3: 5.02 ms per loop

**Comment by agarwal-ashish** [comment]
@tejaslodaya Does tf.RaggedTensor work for you case instead of string encoding ? Another common approach is to pad the ragged data and then loop over it carefully to avoid the paddings.

Also vectorization support for control flow is better now and it can handle tf.while_loop and tf.cond. i.e. the `fn` passed to tf.vectorized_map can have loops and branches. 
See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/parallel_for/control_flow_ops_test.py#L1429
for example that uses an internal `pfor` API for testing.

**Comment by agarwal-ashish** [comment]
@igormorgado A common pitfall with vectorized_map usage is that the call includes the cost of code vectorization in addition to code execution. So it should typically be placed within a tf.function

Example:
def f(x):
  return tf.vectorized_map(lambda z: tf.math.log(z), x)

inp = tf.random.uniform([200, 200])
f(inp)  # warmup
# Slow call since it vectorizes code in each call!
%timeit f(inp)
-> 10 loops, best of 3: 28.9 ms per loop

compiled_f = tf.function(f)
compiled_f(inp)  # warmup
# Vectorization process is done once
%timeit compiled_f(inp)
-> 1000 loops, best of 3: 328 µs per loop

NumPy:
def np_f(x): 
  return np.apply_along_axis(lambda z: np.log(z), 0, arr=x)

np_inp = inp.numpy()
np_f(np_inp)  # warmup
%timeit np_f(np_inp)
-> 1000 loops, best of 3: 961 µs per loop

# Manually vectorized NumPy
np.log(np_inp)  # warmup
%timeit(np.log(np_inp))
-> 1000 loops, best of 3: 249 µs per loop

**Comment by agarwal-ashish** [comment]
@vidursatija 

Here are some benchmarks:

def f(elems):
  return tf.map_fn(lambda x: x**6 , elems, dtype=tf.float64, parallel_iterations=56)

elems = tf.range(1,1000, dtype=np.float64)

f(elems)
%timeit f(elems)
-> 10 loops, best of 3: 162 ms per loop

compiled_f = tf.function(f)
compiled_f(elems)
%timeit compiled_f(elems)
->100 loops, best of 3: 12.8 ms per loop

@tf.function
def vectorized_f(elems):
  return tf.vectorized_map(lambda x: x**6 , elems)

vectorized_f(elems)
%timeit vectorized_f(elems)
->1000 loops, best of 3: 297 µs per loop

**Comment by mhorlacher** [comment]
@agarwal-ashish thanks for the code examples, would you mind formatting it to make it more readable?

**Comment by mhorlacher** [comment]
`tf.map_fn` was slow for me as well, which is why I was switching to `tf.vectorized_map`, however some TensorFlow Ops seem to be not supported / vectorized with `tf.vectorized_map`. 

For example:

```
dataset = tf.data.Dataset.from_tensor_slices([b'\x00\x00\x00\x00']*512)

def f(x):
    x1 = tf.strings.substr(x, 0, 2)
    x2 = tf.strings.substr(x, 2, 2)

    return tf.io.decode_raw(x1, tf.uint8), tf.io.decode_raw(x2, tf.uint8)

dataset = dataset.batch(32).map(lambda x: tf.vectorized_map(f, x))
```

```
WARNING:tensorflow:Using a while_loop for converting Substr
WARNING:tensorflow:Using a while_loop for converting DecodeRaw
WARNING:tensorflow:Using a while_loop for converting Substr
WARNING:tensorflow:Using a while_loop for converting DecodeRaw
```

Which results in execution times of `vectorized_map` being similar to `map_fn`. 

Any ideas to solve this?

**Comment by agarwal-ashish** [comment]
vectorized_map does require registering a "converter" for each op that defines how that op is vectorized. Please see comments for RegisterPFor that describes how these can be defined, in case you want to contribute these.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/parallel_for/pfor.py#L844

**Comment by agarwal-ashish** [comment]
@jsimsa for dataset map calls which has its own vectorization support

**Comment by mhorlacher** [comment]
Thanks @agarwal-ashish, but I fear that registering the a converter for the above ops is beyond my skill set, as I just started with TensorFlow.

**Comment by tejaslodaya** [comment]
@agarwal-ashish in my case, padding it with 0s and carefully parsing them out using raggedtensors would mean padding all the features till the max of length of all rows.


For anyone having the same problem, we overcame this requirement by creating an explicit training-prediction skew with transformed features used for training (till model function) and applying transformations only for prediction.

Considering our models are served on tensorflow serving, we have single, one-off predictions rather than a bulk call. Might not work if your models are offline, batch requests.

**Comment by mohantym** [comment]
Hi @minhg! We see that you are using old version of Tensorflow which is officially considered as end of life, We recommend that you upgrade to 2.6 version and let us know if the issue still persists in newer versions .Please open a new issue in case you face any errors, we will get you the right help .Thanks!
